modules/js-and-css/speculation-rules/hooks.php

```diff
@@ -0,0 +1,66 @@
+<?php
+/**
+ * Hook callbacks used for Speculation Rules.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Prints the speculation rules in a cross-browser compatible way.
+ *
+ * For browsers that do not support speculation rules yet, the rules will not be loaded.
+ *
+ * @since n.e.x.t
+ */
+function plsr_print_speculation_rules() {
+	$rules = plsr_get_speculation_rules();
+	if ( empty( $rules ) ) {
+		return;
+	}
+
+	$script = <<<JS
+( function() {
+	if ( ! HTMLScriptElement.supports || ! HTMLScriptElement.supports( 'speculationrules' ) ) {
+		console.log( 'speculation rules not supported' );
+		return;
+	}
+
+	var specScript = document.createElement( 'script' );
+	specScript.type = 'speculationrules';
+	specScript.textContent = '%s';
+	console.log( 'speculation rules added' );
+	document.body.append( specScript );
+} )();
+JS;
+
+	wp_print_inline_script_tag(
+		sprintf(
+			$script,
+			wp_json_encode( $rules )
```

Comment:
Currently this requires that there be no single quotes in the `$rules` or else a syntax error will occur. This can be guarded against with:

```suggestion
			esc_js( wp_json_encode( $rules ) )
```

However, why not just print the `script[type="speculationrules"]` element directly without adding it dynamically?

Reply:
The JS approach allows to check for browser support first before uncondtionally adding the script tag. It's also referenced in the Chrome Developer blog article.

Reply:
Curious why that approach was taken in the article. If the browser doesn't support it, it would just be ignored, right? It would seem preferable to just do:

```php
wp_print_inline_script_tag(
    wp_json_encode( $rules ),
    array( 'type' => 'speculationrules' )
);
```

Reply:
Potentially yes. Let's sort out whether it doesn't have any side effects on browsers that don't support it - if so, I agree it's simplest to just include like you're saying.